### PR TITLE
Add content filters to repository endpoint

### LINKED ISSUE #2865: As a user, I want to know which repositories some content is in
**Is your feature request related to a problem? Please describe.**
Content in pulp is distributed through repositories and distributions. If a client knows the name or HREF for some content, there's currently no way of know which repositories it's available in without looking in each repository. 

**Describe the solution you'd like**
Add a `?content_in_latest=` filter to the distributions and repositories APIs that allow users to specify a content HREF and get a list of repositories or distributions that contain the selected content.

**Describe alternatives you've considered**
Are there any existing ways to do this in pulp?

**Additional context**
This is needed for repository management in galaxy_ng

### PR DESCRIPTION
**Pull request by daviddavis**
fixes #2865

### CODE REVIEW

**Comment by daviddavis** [comment]
I can add tests if this change is acceptable.

File: pulpcore/app/viewsets/repository.py [hunk 1]
```diff
@@ -41,9 +42,51 @@
...
+                ).values_list("repository__pk", flat=True)
+            )
+        else:
+            return qs.filter(content__pk=content.pk)
+
+
 class RepositoryFilter(BaseFilterSet):
     pulp_label_select = LabelFilter()
     remote = HyperlinkRelatedFilter(allow_null=True)
+    with_content = RepositoryContentFilter()
```
Comment (daviddavis) [inline pulpcore/app/viewsets/repository.py:88]: I wasn't sure what to name this filter but it seems like [other places used the name `with_content`](https://github.com/pulp/pulpcore/blob/765c70aabb69a6ea71d2841d19d3535cf9dd412e/pulpcore/app/viewsets/publication.py#L364).

File: pulpcore/app/viewsets/repository.py [hunk 2]
```diff
@@ -41,9 +42,51 @@
...
+    currently_with_content = RepositoryContentFilter(latest=True)
```
Comment (mdellweg) [inline pulpcore/app/viewsets/repository.py:89]: Should we name this "latest_with_content"? Maybe easier to relate to the "latest_version"?

Being exposed by distributions, a repository can have multiple current versions...

File: pulpcore/app/viewsets/repository.py [hunk 3]
```diff
@@ -41,9 +42,51 @@
...
+
+        if value is None:
+            # user didn't supply a value
+            return qs
+
+        if not value:
+            raise serializers.ValidationError(detail=_("No value supplied for content filter"))
+
+        # Get the content object from the content_href
+        content = NamedModelViewSet.get_resource(value, Content)
```
Comment (mdellweg) [inline pulpcore/app/viewsets/repository.py:73]: This, I'd expect to throw 404s, we should translate to ValidationError too.

File: pulpcore/app/viewsets/repository.py (hunk 1 above)
Comment (mdellweg) [reply pulpcore/app/viewsets/repository.py:88]: Yes, i think this name is fine.

File: pulpcore/app/viewsets/repository.py (hunk 3 above)
Comment (daviddavis) [reply pulpcore/app/viewsets/repository.py:73]: The method throws validation errors already depending on the problem. Here's an example:

https://github.com/pulp/pulpcore/blob/8b665bb00ad10d14d745f256017779ff38dbf6fb/pulpcore/app/viewsets/base.py#L193

**Comment by daviddavis** [comment]
Test PR: https://github.com/pulp/pulp_file/pull/953

File: pulpcore/app/viewsets/repository.py [hunk 4]
```diff
@@ -41,9 +42,51 @@
...
+
+        if self.latest:
+            return qs.filter(
+                pk__in=RepositoryContent.objects.filter(
+                    version_removed=None, content__pk=content.pk
+                ).values_list("repository__pk", flat=True)
+            )
+        else:
+            return qs.filter(content__pk=content.pk)
```
Comment (lubosmj) [inline pulpcore/app/viewsets/repository.py:82]: What do we do about the "user_hidden" field?

File: pulpcore/app/viewsets/repository.py [hunk 5]
```diff
@@ -41,9 +42,51 @@
...
+
+        # Get the content object from the content_href
+        content = NamedModelViewSet.get_resource(value, Content)
+
+        if self.latest:
+            return qs.filter(
+                pk__in=RepositoryContent.objects.filter(
+                    version_removed=None, content__pk=content.pk
+                ).values_list("repository__pk", flat=True)
+            )
```
Comment (lubosmj) [inline pulpcore/app/viewsets/repository.py:80]: Will this work when we set retain_repo_versions to 1?

https://docs.pulpproject.org/pulpcore/workflows/repo-versioning.html#version-retention

**Review by lubosmj — COMMENTED** [review]
@ipanova was the one who did not embrace the need of creating such a filter before. Can you add your thoughts here, @ipanova? I like the change.

File: pulpcore/app/viewsets/repository.py (hunk 5 above)
Comment (daviddavis) [reply pulpcore/app/viewsets/repository.py:80]: Yes, it should. Setting retain_repo_versions to 1 should just keep the latest version (where version_removed=None).

**Comment by daviddavis** [comment]
@lubosmj @ipanova the other approach that was proposed in the issue was adding a `latest` filter to the repo version endpoint. I think it's a great idea but it wouldn't work in our particular use case. Our users only deal with repos: they create, sync, modify, publish, and distribute repos. So we don't expose repo versions or the repo version endpoints to our users and therefore users can't search for repos based on content.

**Review by mdellweg — APPROVED** [review]

**Comment by mdellweg** [comment]
Last pulpcore meeting we had a discussion that new test (like these) that only test a new pulpcore feature without any necessary adoption in the plugin should come with the code and live in this repository. Are you willing to move your's here? (This is not for existing tests, because it does not really matter for them.)

**Comment by daviddavis** [comment]
@mdellweg ok, done.

File: pulpcore/tests/functional/api/test_repos.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    monitor_task(response.task)
+    repo = file_repository_api_client.read(repo.pulp_href)
+
+    # the repo still has the content unit
+    results = file_repository_api_client.list(with_content=content.pulp_href).results
+    assert [repo] == results
+
+    # but not in its latest version anymore
+    results = file_repository_api_client.list(latest_with_content=content.pulp_href).results
+    assert [] == results
```
Comment (mdellweg) [inline pulpcore/tests/functional/api/test_repos.py:50]: I prefer to have it "assert result == expected". Just a personal thing probably, so feel free to ignore.

File: pulpcore/tests/functional/api/test_repos.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    file_repository_api_client,
+    file_remote_factory,
+    gen_object_with_cleanup,
+    write_3_iso_file_fixture_data_factory,
+    monitor_task,
+):
+    """Test repository's content filters."""
+    # generate a repo with some content
+    repo = gen_object_with_cleanup(file_repository_api_client, gen_repo())
+    repo_manifest_path = write_3_iso_file_fixture_data_factory(utils.uuid4())
```
Comment (mdellweg) [inline pulpcore/tests/functional/api/test_repos.py:22]: Let's try not to use any of the pulp-smash weirdness. There should be a file_repository_factory and we can use `str(uuid.uuid4())`.

**Review by mdellweg — APPROVED** [review]

**Pull request merged**